Split point cloud index into implementation & wrapper

dvdkon opened the pull request:
## Description

This PR renames `QgsPointCloudIndex` to `QgsAbstractPointCloudIndex` and introduces a new class, `QgsPointCloudIndex`, which serves as a wrapper for a `shared_ptr` of the abstract index. This mirrors the arrangement used for `QgsTiledSceneIndex` and codifies good memory management practices (use of smart pointers for a long-lived shared object) while not exposing the `std::shared_pointer` directly in the API.

Personally, I see the use of a wrapper class as a hack around SIP's lack of support for smart pointers and would much rather see a simple `using QgsPointCloudIndex = std::shared_ptr<...>;`. That would, however, require much more complex changes, and the imperfect route chosen in this PR allows us to use good memory management practices today, without any long-term debt (in the future it would be simple to switch to using smart pointers directly).

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

dvdkon committed:
Split point cloud index into implementation & wrapper

nyalldawson commented on src/3d/qgspointcloudlayerchunkloader_p.h:
```diff
@@ -63,7 +63,7 @@ class QgsPointCloudLayerChunkLoaderFactory : public QgsChunkLoaderFactory
      * The factory takes ownership over the passed \a symbol
      */
-    QgsPointCloudLayerChunkLoaderFactory( const Qgs3DRenderContext &context, const QgsCoordinateTransform &coordinateTransform, QgsPointCloudIndex *pc, QgsPointCloud3DSymbol *symbol, double zValueScale, double zValueOffset, int pointBudget );
+    QgsPointCloudLayerChunkLoaderFactory( const Qgs3DRenderContext &context, const QgsCoordinateTransform &coordinateTransform, QgsPointCloudIndex pc, QgsPointCloud3DSymbol *symbol, double zValueScale, double zValueOffset, int pointBudget );
```
```suggestion
    QgsPointCloudLayerChunkLoaderFactory( const Qgs3DRenderContext &context, const QgsCoordinateTransform &coordinateTransform, const QgsPointCloudIndex &pc, QgsPointCloud3DSymbol *symbol, double zValueScale, double zValueOffset, int pointBudget );
```

nyalldawson commented on src/3d/qgspointcloudlayerchunkloader_p.h:
```diff
@@ -127,7 +127,7 @@ class QgsPointCloudLayerChunkedEntity : public QgsChunkedEntity
     Q_OBJECT
   public:
-    explicit QgsPointCloudLayerChunkedEntity( Qgs3DMapSettings *map, QgsPointCloudIndex *pc, const QgsCoordinateTransform &coordinateTransform, QgsPointCloud3DSymbol *symbol, float maxScreenError, bool showBoundingBoxes, double zValueScale, double zValueOffset, int pointBudget );
+    explicit QgsPointCloudLayerChunkedEntity( Qgs3DMapSettings *map, QgsPointCloudIndex pc, const QgsCoordinateTransform &coordinateTransform, QgsPointCloud3DSymbol *symbol, float maxScreenError, bool showBoundingBoxes, double zValueScale, double zValueOffset, int pointBudget );
```
```suggestion
    explicit QgsPointCloudLayerChunkedEntity( Qgs3DMapSettings *map, const QgsPointCloudIndex &pc, const QgsCoordinateTransform &coordinateTransform, QgsPointCloud3DSymbol *symbol, float maxScreenError, bool showBoundingBoxes, double zValueScale, double zValueOffset, int pointBudget );
```

nyalldawson commented on src/core/pointcloud/qgspointcloudindex.h:
```diff
@@ -197,35 +198,38 @@ class CORE_EXPORT QgsPointCloudNode
 };
 
+//! The access type of the data, local is for local files and remote for remote files (over HTTP)
+enum class QgsPointCloudAccessType
```
Move to Qgis, rename to PointCloudAccessType

nyalldawson commented on src/3d/symbols/qgspointcloud3dsymbol_p.h:
```diff
@@ -76,14 +76,14 @@ class QgsPointCloud3DSymbolHandler
     virtual Qt3DCore::QGeometry *makeGeometry( Qt3DCore::QNode *parent, const QgsPointCloud3DSymbolHandler::PointData &data, unsigned int byteStride ) = 0;
 #endif
-    std::unique_ptr<QgsPointCloudBlock> pointCloudBlock( QgsPointCloudIndex *pc, const QgsPointCloudNodeId &node, const QgsPointCloudRequest &request, const QgsPointCloud3DRenderContext &context );
+    std::unique_ptr<QgsPointCloudBlock> pointCloudBlock( QgsPointCloudIndex pc, const QgsPointCloudNodeId &node, const QgsPointCloudRequest &request, const QgsPointCloud3DRenderContext &context );
```
```suggestion
    std::unique_ptr<QgsPointCloudBlock> pointCloudBlock( const QgsPointCloudIndex &pc, const QgsPointCloudNodeId &node, const QgsPointCloudRequest &request, const QgsPointCloud3DRenderContext &context );
```

elsewhere too

nyalldawson commented on src/core/pointcloud/qgscopcpointcloudindex.h:
```diff
@@ -84,11 +84,11 @@ class CORE_EXPORT QgsCopcPointCloudIndex: public QgsPointCloudIndex
-     * \since QGIS 3.42
      */
-    bool gpsTimeFlag() const;
+    QVariantMap extraMetadata() const override;
```
This is more fragile vs the explicit getter -- I'd rather not see us use QVariantMap for purposes like this

wonder-sk replied:
@nyalldawson the reasoning is that the gpsTimeFlag is something very specific to one particular file format, with quite minor impact, and it does not deserve to be a part of the public API. And it seemed that over time, we may want to pass more format-specific metadata read from the file header - hence introduction of this fairly generic extraMetadata() function. Given that it is not used for any core functionality, it seems okay to use QVariantMap in this case (I also don't like to see QVariantMap in API)

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -381,11 +387,11 @@ bool QgsPointCloudDataProvider::setSubsetString( const QString &subset, bool upd
 
   Q_UNUSED( updateFeatureCount )
-  const auto i = index();
+  auto i = index();
```
use explicit type

nyalldawson commented on src/core/pointcloud/qgspointcloudindex.h:
```diff
@@ -378,4 +397,243 @@ class CORE_EXPORT QgsPointCloudIndex
+{
+  public:
+    //! Construct wrapper, takes ownership of index
+    QgsPointCloudIndex( QgsAbstractPointCloudIndex *index ) SIP_SKIP;
```
```suggestion
    explicit QgsPointCloudIndex( QgsAbstractPointCloudIndex *index ) SIP_SKIP;
```

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.h:
```diff
@@ -157,7 +157,7 @@ class CORE_EXPORT QgsPointCloudDataProvider: public QgsDataProvider
      * \note Not available in Python bindings
      */
-    virtual QgsPointCloudIndex *index() const SIP_SKIP {return nullptr;}
+    virtual QgsPointCloudIndex index() const SIP_SKIP {return nullptr;}
```
Let's make this a little clearer:
```suggestion
    virtual QgsPointCloudIndex index() const SIP_SKIP {return QgsPointCloudIndex( nullptr );}
```

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.h:
```diff
@@ -279,10 +279,10 @@ class CORE_EXPORT QgsPointCloudDataProvider: public QgsDataProvider
 
     //! Identify in a specific index (used for sub-indexes)
-    QVector<QVariantMap> identify( QgsPointCloudIndex *index, double maxError, const QgsGeometry &extentGeometry, const QgsDoubleRange &extentZRange, int pointsLimit ) SIP_SKIP ;
+    QVector<QVariantMap> identify( QgsPointCloudIndex index, double maxError, const QgsGeometry &extentGeometry, const QgsDoubleRange &extentZRange, int pointsLimit ) SIP_SKIP ;
```
```suggestion
    QVector<QVariantMap> identify( const QgsPointCloudIndex &index, double maxError, const QgsGeometry &extentGeometry, const QgsDoubleRange &extentZRange, int pointsLimit ) SIP_SKIP ;
```
elsewhere too

dvdkon replied:
Thanks for the comments, I'll look through them tomorrow. But I don't think `QgsPointCloudIndex` should be passed by reference. The point of the class is that it's basically an `std::shared_ptr`. Passing it by reference might remove a refcount increment at the cost of more indirection, but I'd rather just pass it by value for simplicity.

dvdkon replied:
If we had the full range of C++ to play with, I'd personally either use `std::shared_ptr<QgsAbstractPointCloudIndex>` or `QgsAbstractPointCloudIndex &`, but as-is I'd just use the wrapper everywhere.

nyalldawson replied:
Hmm -- I would still push towards references, in all cases EXCEPT those where we know that we're explicitly storing the result in a member variable (and could possibly benefit from a move).

(In general we try to follow Qt's approach for things like this, and they **always** pass QString by reference, even though it's implicitly shared and copies are cheap.)

nyalldawson commented:
Looking good, thank you! Can you also add some python based tests just to ensure that the bindings are working correctly?

dvdkon committed:
Move QgsPointCloudAccessType to Qgis

dvdkon committed:
Apply changes per review

dvdkon committed:
Expose QgsPointCloudIndex to Python

dvdkon commented:
> Looking good, thank you! Can you also add some python based tests just to ensure that the bindings are working correctly?

Thanks! I've exposed the `index()` method on the data provider as well and added a simple Python test accessing the index.

I've had some trouble getting the SIP bindings to build after this one change, hopefully it passes CI.

dvdkon committed:
Fix SIP build errors by using /NoTypeName/

The typedef is only present in the .sip file, but without this
annotation SIP assumes it also exists in QGIS headers. This didn't
completely break the build, since we concatenate multiple SIP headers
into larger files, so the typedefs from one part "fixed" another. Run
sip-build without --concatenate to see the issue clearly.

This fix is currently sadly not fully effective due to a SIP bug:
https://github.com/Python-SIP/sip/issues/66

dvdkon committed:
Work around SIP bug

dvdkon commented:
This PR ran into a long-standing QGIS build issue, where if the SIP files weren't concatenated "right", missing `typedefs` would break the build. This was because the typedefs used in .sip files for inheriting from template classes ([not possible directly](https://github.com/Python-SIP/sip/issues/12)) weren't annotated correctly with `/NoTypeName/`, so SIP assumed that the typedef was also present in QGIS headers. The concatenation process papered over this by letting a typedef from one file (SIP creates a file per class by default) affect another.

I fixed this issue in d586c28c74436d82b1bd61027ab38fe7c0437ff9, but also had to put in a quick bodge in 486e3b0f4d56ddcc4596a58dca9e766cd7f642d6 to work around Python-SIP/sip#66. I've added the commits to this PR, since they're necessary to consistently build it.

wonder-sk reviewed (approved):
Looking great, thanks!

And thanks a lot for looking into the sip build issues! :clap:
